### PR TITLE
[#20] Add Expiration

### PR DESCRIPTION
**Pull request by eubnara**
- add option "stdTTL" as seconds
- only ttl per namespace supported
  - cannot specify ttl per key
- not affect to "getAll()"

### CODE REVIEW

**Comment by wsdt** [comment]
Looks good on first sight 👌

File: src/cache.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ export interface ICacheOptions {
     backend: any;
     namespace: string;
     policy: ICachePolicy;
+    stdTTL: number; // second
```
Comment (jstrunk) [inline src/cache.ts:6]: Why not make this part of the `policy`?

File: src/cache.ts [hunk 2]
```diff
@@ -13,17 +14,23 @@ export default class Cache {
...
+            ttl = 0;
+        }
+        this.stdTTL = ttl;
     }
 
     public async clearAll() {
         const keys = await this.backend.getAllKeys();
         const namespaceKeys = keys.filter((key: string) => {
-            return key.substr(0, this.namespace.length) === this.namespace;
+            return key.split(':')[0] === this.namespace;
```
Comment (jstrunk) [inline src/cache.ts:33]: This looks unrelated. Should it be part of a different PR?

File: src/cache.ts [hunk 3]
```diff
@@ -54,7 +61,7 @@ export default class Cache {
     public async getAll() {
         const keys = await this.backend.getAllKeys();
         const namespaceKeys = keys.filter((key: string) => {
-            return key.substr(0, this.namespace.length) === this.namespace;
+            return key.split(':')[0] === this.namespace;
```
Comment (jstrunk) [inline src/cache.ts:64]: This looks unrelated. Should it be part of a different PR?

File: src/cache.ts [hunk 4]
```diff
@@ -102,6 +109,14 @@ export default class Cache {
         let value;
         if (entry) {
             value = entry.value;
+            if (this.stdTTL > 0) {
+                const deadline = entry.created.getTime() + this.stdTTL * 1000;
+                const now = (new Date()).getTime();
+                if (deadline < now) {
+                    await this.remove(key);
```
Comment (jstrunk) [inline src/cache.ts:116]: Wouldn't it be better to not wait for removal? If it's a cache miss, we want to go to the origin right away.

**Comment by eubnara** [comment]
@jstrunk 
Thanks for your review.
Sorry for being late. I changed it according to your review.

**Comment by timfpark** [comment]
Sorry - one more - can you add how this works to the documentation as well?

File: src/cache.ts [hunk 5]
```diff
@@ -102,6 +108,14 @@ export default class Cache {
         let value;
         if (entry) {
             value = entry.value;
+            if (this.policy.stdTTL > 0) {
+                const deadline = entry.created.getTime() + this.policy.stdTTL * 1000;
+                const now = (new Date()).getTime();
```
Comment (timfpark) [inline src/cache.ts:113]: Instead of (new Date()).getTime() - can you use Date.now() in the if statement below which is functionally the same.  Otherwise, this looks great - thank you for the enhancement!

**Review by timfpark — COMMENTED** [review]
Looks great - can you fix the one comment and then also add a line with how this works to the README.md - thank you for the contribution.

**Comment by eubnara** [comment]
@timfpark 
I'm sorry for delay.
Thanks for review again. I changed some codes according to your feedback.

**Review by timfpark — APPROVED** [review]
Looks great - thank you for the contribution!
